sycl/include/CL/sycl/detail/context_impl.hpp

```diff
@@ -58,14 +58,22 @@ class context_impl {
   RT::PiContext &getHandleRef();
   const RT::PiContext &getHandleRef() const;
 
+  // Operating the returned map won't be thread-safe,
+  // use getProgramKernelCacheMutex() along with this method
   std::map<OSModuleHandle, RT::PiProgram> &getCachedPrograms() {
     return m_CachedPrograms;
   }
+  // Operating the returned map won't be thread-safe,
+  // use getProgramKernelCacheMutex() along with this method
```

Comment:
(Disclaimer: I'm far from competent in the RT/thread safety)
Is it really a good idea to rely on the method user instead of wrapping the `std::map` and the mutex object into a separate class and re-defining methods to lock properly?

Reply:
> instead of wrapping the std::map and the mutex object into a separate class and re-defining the methods to lock properly?

This is easy-to-use thing, but it might be non-performant one: sometimes you are already under a lock and it is not necessary to acquire additional lock for **each** operation which is done on the map.

> Is it really a good idea to rely on the method user

This one assumes responsibility of a developer and at the same time allows achieve better performance by omitting locks where unnecessary or grouping several operations with map under the same lock to reduce overheads to acquire/release mutex.